src/mds/CInode.cc

```diff
@@ -2112,6 +2117,9 @@ void CInode::decode_lock_ixattr(bufferlist::const_iterator& p)
   }
   DECODE_FINISH(p);
   reset_inode(std::move(_inode));
+
+  std::string bal_rank_mask = get_bal_rank_mask_from_xattrs(false);
+  maybe_export_pin(prev_bal_rank_mask != bal_rank_mask);
```

Comment:
Let's accept this as-is, since we asked you to do it this way; but, I plan to restructure this to use the policylock (which is the right lock). I don't like putting this information in the xattrmap as it relies on using the cap-related locks and will introduce numerous similar checks for every ixattr decode. It's also simply incoherent with the current design of locks.

My plan is to create another lightweight map in the CInode which can take any number of these rarely set metadata values so it's efficient to encode/decode. The in-memory version would be a std::vector for efficient lookup (a map is unnecessary for small, and often empty, datasets).

If you want to make that change before merging this please let me know. Otherwise, I'll do it post-merge.

Reply:
I think developing a new lightweight map for CInode looks very interesting. However, it would be better to proceed separately from this PR. I prefer to merge this PR first.